### PR TITLE
Split editor to x-plat and Windows (#3519)

### PR DESCRIPTION
**Pull request by AlexanderSher**
From @MikhailArkhipov :
The change is huge.

I am doing manual testing and unit tests pass except some that need editor MEF at the package level. I am
working with IDE folks on getting correct nugets. 15.3 dogfood is required. Most probably PR will receive few more changes depending on the package level test outcomes.

Basically changes amount to

- abstracting text buffer via IEditorBuffer and bunch of friends
- abstracting view via IEditorView
- abstracting settings storage access
- getting rid of most statics
- cleaning old code, moving bunch of methods to extensions
- using text view/buffer singletons for VS-specific services and IEditorBuffer/View services for xplat

### CODE REVIEW

File: src/Common/Core/Impl/Collections/EnumerableExtensions.cs [hunk 1]
```diff
@@ -10,12 +10,12 @@ public static class EnumerableExtensions {
         public static List<T> AsList<T>(this IEnumerable<T> source) => source as List<T> ?? source.ToList();
         public static T[] AsArray<T>(this IEnumerable<T> source) => source as T[] ?? source.ToArray();
 
-        public static IEnumerable<T> Append<T>(this IEnumerable<T> source, T item) {
-            foreach (T sourceItem in source) {
-                yield return sourceItem;
-            }
-            yield return item;
-        }
+        //public static IEnumerable<T> Append<T>(this IEnumerable<T> source, T item) {
```
Comment (AlexanderSher) [inline src/Common/Core/Impl/Collections/EnumerableExtensions.cs]: Why? There is any problem with this method?

File: src/Common/Core/Impl/Diagnostics/Check.cs [hunk 2]
```diff
@@ -19,16 +22,25 @@ public static class Check {
...
         public static void InvalidOperation(Func<bool> predicate, string message = null) {
             if (!predicate()) {
                 throw new InvalidOperationException(message ?? string.Empty);
             }
         }
+
+        [DebuggerStepThrough]
+        public static void Argument(string argumentName, Func<bool> predicate) {
+            if (!predicate()) {
+                throw new ArgumentException();
```
Comment (AlexanderSher) [inline src/Common/Core/Impl/Diagnostics/Check.cs]: `ArgumentException($"{argumentName} is not valid", argumentName)` or something like that

File: src/Common/Core/Impl/Extensions/ExceptionExtensions.cs [hunk 3]
```diff
@@ -9,16 +9,7 @@ public static class ExceptionExtensions {
...
-#if NETSTANDARD1_6
-            return ex is OutOfMemoryException;
-#else
-            return ex is StackOverflowException 
-                || ex is OutOfMemoryException 
-                || ex is ThreadAbortException 
-                || ex is AccessViolationException;
-#endif
-        }
+        public static bool IsCriticalException(this Exception ex) => ex is OutOfMemoryException;
```
Comment (AlexanderSher) [inline src/Common/Core/Impl/Extensions/ExceptionExtensions.cs]: In NETSTANDARD2 these exceptions should become available.

File: src/Common/Core/Impl/Extensions/IOExtensions.cs [hunk 4]
```diff
@@ -63,6 +59,28 @@ public static class IOExtensions {
             return files;
         }
 
+        /// <summary>
+        /// Returns subdirectories of a directory
+        /// </summary>
+        public static IEnumerable<string> GetDirectories(this IFileSystem fs, string directory) {
+            foreach (var fsi in fs.GetDirectoryInfo(directory).EnumerateFileSystemInfos()) {
```
Comment (AlexanderSher) [inline src/Common/Core/Impl/Extensions/IOExtensions.cs]:
```cs
=> fs.GetDirectoryInfo(directory)
    .EnumerateFileSystemInfos()
    .Where(fsi => (fsi.Attributes & FileAttributes.Directory) != 0)
    .Select(fsi => fsi.FullName);
```

Comment (AlexanderSher) [inline src/Common/Core/Impl/Extensions/IOExtensions.cs]: Btw, I think we can return `IEnumerable<IFileSystemInfo>`

File: src/Common/Core/Impl/Idle/CancellableTask.cs [hunk 5]
```diff
@@ -18,33 +18,30 @@ namespace Microsoft.Languages.Editor.Tasks {
     // from background thread directly rather than making Dispatcher.Invoke call
     // that will try transitioning data to the main thread.
     public abstract class CancellableTask : IDisposable {
-        private Task _task = null;
-        private ManualResetEventSlim _taskCompleted = new ManualResetEventSlim(true);
-        private long _canceled = 0;
+        // TODO: clean old code, use cancellation token, etc
```
Comment (AlexanderSher) [inline src/Common/Core/Impl/Idle/CancellableTask.cs]: Do you want to add a task for it?

File: src/Common/Core/Impl/Idle/IdleTimeAsyncTask.cs [hunk 6]
```diff
@@ -26,8 +31,10 @@ public sealed class IdleTimeAsyncTask : IDisposable {
 
         public object Tag { get; private set; }
 
-        public IdleTimeAsyncTask(ICoreShell shell) {
-            _shell = shell;
+        public IdleTimeAsyncTask(IServiceContainer services) {
+            _idleTime = services.GetService<IIdleTimeService>();
+            _mainThread = services.GetService<IMainThread>();
```
Comment (AlexanderSher) [inline src/Common/Core/Impl/Idle/IdleTimeAsyncTask.cs]: `services.MainThread()`

File: src/Common/Core/Impl/Security/Credentials.cs [hunk 7]
```diff
@@ -18,12 +18,6 @@ public sealed class Credentials : ICredentials {
...
         }
 
-        public NetworkCredential GetCredential(Uri uri, string authType) {
-#if NETSTANDARD1_6
-            return new NetworkCredential(UserName, Password.ToUnsecureString());
-#else
-            return new NetworkCredential(UserName, Password);
-#endif            
-        }
+        public NetworkCredential GetCredential(Uri uri, string authType) => new NetworkCredential(UserName, Password.ToUnsecureString());
```
Comment (AlexanderSher) [inline src/Common/Core/Impl/Security/Credentials.cs]: In `NETSTANDARD2` there will be constructor with `SecureString` - let's not remove this condition.

File: src/Common/Core/Impl/Services/ServiceManager.cs [hunk 8]
```diff
@@ -40,24 +33,36 @@ public class ServiceManager : IServiceManager {
...
+            var lazy = new Lazy<object>(() => factory(this));
+            Check.InvalidOperation(() => _s.TryAdd(typeof(T), lazy), "Service already exists");
+            return this;
+        }
+
+        /// <summary>
+        /// Adds on-demand created service
+        /// </summary>
+        /// <param name="type">Type to instantiate with reflection.</param>
+        public virtual IServiceManager AddService(Type type) {
```
Comment (AlexanderSher) [inline src/Common/Core/Impl/Services/ServiceManager.cs]: Why `AddService<T>()` didn't work?

File: src/Common/Core/Impl/Services/ServiceManager.cs [hunk 9]
```diff
@@ -40,24 +33,36 @@ public class ServiceManager : IServiceManager {
             type = type ?? typeof(T);
             Check.ArgumentNull(nameof(service), service);
             Check.InvalidOperation(() => _s.TryAdd(type, service), "Service already exists");
-            ServiceAdded?.Invoke(this, new ServiceContainerEventArgs(type));
+
+            var mainThread = service as IMainThread;
```
Comment (AlexanderSher) [inline src/Common/Core/Impl/Services/ServiceManager.cs]: Why `IMainThread` is special-cased?

File: src/Common/Core/Test/Fixtures/ServiceManagerWithMefFixture.cs [hunk 10]
```diff
@@ -33,6 +33,7 @@ public abstract class ServiceManagerWithMefFixture : ServiceManagerFixture {
 
         private sealed class TestServiceManagerWithMef : TestServiceManager {
             private readonly CompositionContainer _compositionContainer;
+            private volatile bool _disposed;
```
Comment (AlexanderSher) [inline src/Common/Core/Test/Fixtures/ServiceManagerWithMefFixture.cs]: You can use `DisposableToken` here.

File: src/Common/Core/Impl/Services/ServiceManager.cs [hunk 11]
```diff
@@ -77,25 +92,21 @@ public class ServiceManager : IServiceManager {
             return (T)CheckDisposed(value as T ?? (value as Lazy<object>)?.Value);
         }
 
-        public virtual void RemoveService<T>() where T : class {
-            _disposeToken.ThrowIfDisposed();
-
-            if (_s.TryRemove(typeof(T), out object dummy)) {
-                ServiceRemoved?.Invoke(this, new ServiceContainerEventArgs(typeof(T)));
+        public virtual void RemoveService(object service) {
```
Comment (AlexanderSher) [inline src/Common/Core/Impl/Services/ServiceManager.cs]: Why `_s.TryRemove(service.GetType())` doesn't work?

**Comment by AlexanderSher** [comment]
Btw, VS packages with version 15.0.26507.55 aren't available in default nuget feed. What feeds do you have?

**Comment by MikhailArkhipov** [comment]
See OneNote

```
https://www.myget.org/F/vs-devcore/api/v3/index.json (VS 2017+)
https://www.myget.org/F/vs-editor/api/v3/index.json
https://dotnet.myget.org/F/interactive-window/api/v3/index.json
```

**Comment by AlexanderSher** [comment]
Ok, I'll add it to the config

**Comment by MikhailArkhipov** [comment]
I am actually going to try 15.2 instead. I still don't have matching editor components for 15.3

**Comment by AlexanderSher** [comment]
Ok, build have passed for everything except Package. Do I need to update VSSDK?

**Comment by MikhailArkhipov** [comment]
I can't seem to match CPS and IDE/Editor bits for 15.3 :-( - MEF doesn't like one or another depending on the combination. I have CPS for 15.2 now and will try that.

File: src/Common/Core/Impl/Collections/EnumerableExtensions.cs (hunk 1 above)
Comment (MikhailArkhipov) [inline src/Common/Core/Impl/Collections/EnumerableExtensions.cs]: It conflicts with another Append apparently available from Linq on 4.6 (but not on net core)

File: src/Common/Core/Impl/Services/ServiceManager.cs (hunk 9 above)
Comment (MikhailArkhipov) [inline src/Common/Core/Impl/Services/ServiceManager.cs]: There is code that wants main thread *after* SM is disposed... Mostly tests though (fixture gets disposed BEFORE test actually completes) but I guess there could be cases (idle, async) when call completes after disposal.

File: src/Common/Core/Impl/Services/ServiceManager.cs (hunk 8 above)
Comment (MikhailArkhipov) [inline src/Common/Core/Impl/Services/ServiceManager.cs]: I can probably yank type here, only VS globals use different types.

File: src/Common/Core/Impl/Services/ServiceManager.cs (hunk 9 above)
Comment (AlexanderSher) [inline src/Common/Core/Impl/Services/ServiceManager.cs]: That's not right. Fixtures are disposed inside test pipeline, so if test isn't fully completed at that time, we may get incorrect results: https://github.com/Microsoft/RTVS/issues/3523

File: src/Common/Core/Impl/Services/ServiceManager.cs (hunk 8 above)
Comment (MikhailArkhipov) [inline src/Common/Core/Impl/Services/ServiceManager.cs]: Actually it is needed for test container that adds mocks for VS global services and needs both instance and different type

File: src/Common/Core/Impl/Services/ServiceManager.cs (hunk 9 above)
Comment (MikhailArkhipov) [inline src/Common/Core/Impl/Services/ServiceManager.cs]: Ex, run history tests
```
System.ArgumentNullException: Value cannot be null.
Parameter name: mainThread
   at Microsoft.Common.Core.Diagnostics.Check.ArgumentNull(String argumentName, Object argument) in F:\RTVS\src\Common\Core\Impl\Diagnostics\Check.cs:line 12
   at Microsoft.Common.Core.Threading.MainThreadAwaiter..ctor(IMainThread mainThread, CancellationToken cancellationToken) in F:\RTVS\src\Common\Core\Impl\Threading\MainThreadAwaiter.cs:line 16
   at Microsoft.Common.Core.Threading.MainThreadAwaitable.GetAwaiter() in F:\RTVS\src\Common\Core\Impl\Threading\MainThreadAwaitable.cs:line 16
   at Microsoft.R.Components.InteractiveWorkflow.Implementation.InteractiveWindowConsole.<WriteAsync>d__5.MoveNext() in F:\RTVS\src\Windows\R\Components\Impl\InteractiveWorkflow\Implementation\InteractiveWindowConsole.cs:line 30
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.Common.Core.TaskExtensions.ReThrowTaskException(Object state) in F:\RTVS\src\Common\Core\Impl\Extensions\TaskExtensions.cs:line 78
   at Microsoft.Common.Core.TaskExtensions.DoNotWait(Task task) in F:\RTVS\src\Common\Core\Impl\Extensions\TaskExtensions.cs:line 62
   at Microsoft.R.Components.InteractiveWorkflow.Implementation.InteractiveWindowConsole.WriteError(String text) in F:\RTVS\src\Windows\R\Components\Impl\InteractiveWorkflow\Implementation\InteractiveWindowConsole.cs:line 22
   at Microsoft.R.Components.InteractiveWorkflow.Implementation.InteractiveWindowConsole.WriteErrorLine(String text) in F:\RTVS\src\Windows\R\Components\Impl\InteractiveWorkflow\Implementation\InteractiveWindowConsole.cs:line 42
   at Microsoft.R.Components.InteractiveWorkflow.Implementation.RInteractiveEvaluator.WriteErrorLine(String message) in F:\RTVS\src\Windows\R\Components\Impl\InteractiveWorkflow\Implementation\RInteractiveEvaluator.cs:line 291
   at Microsoft.R.Components.InteractiveWorkflow.Implementation.RInteractiveEvaluator.SessionOnDisconnected(Object sender, EventArgs args) in F:\RTVS\src\Windows\R\Components\Impl\InteractiveWorkflow\Implementation\RInteractiveEvaluator.cs:line 236
   at System.EventHandler`1.Invoke(Object sender, TEventArgs e)
   at Microsoft.R.Host.Client.Session.RSession.Microsoft.R.Host.Client.IRCallbacks.Disconnected() in F:\RTVS\src\Host\Client\Impl\Session\RSession.cs:line 524
   at Microsoft.R.Host.Client.RHost.<RunWorker>d__53.MoveNext() in F:\RTVS\src\Host\Client\Impl\Host\RHost.cs:line 684
--- End of stack trace from previous location where exception was thrown ---

```

File: src/Common/Core/Impl/Services/ServiceManager.cs (hunk 8 above)
Comment (AlexanderSher) [inline src/Common/Core/Impl/Services/ServiceManager.cs]: I thought in that case we use `IServiceManager AddService<T>(T service, Type type = null) where T : class;`
I've found only one usage of `AddService(Type type)` in `ServiceManagerTests`.

File: src/Common/Core/Test/Fixtures/ServiceManagerWithMefFixture.cs [hunk 12]
```diff
@@ -53,17 +54,17 @@ private sealed class TestServiceManagerWithMef : TestServiceManager {
             }
 
             public override Task DisposeAsync(RunSummary result, IMessageBus messageBus) {
-                if (_disposed) {
+                if (_disposeToken.IsDisposed) {
```
Comment (AlexanderSher) [inline src/Common/Core/Test/Fixtures/ServiceManagerWithMefFixture.cs]:
```cs
if (!_disposeToken.TryMarkDisposed()) {
    return Task.CompletedTask;
}
```

File: src/Common/Core/Impl/Services/ServiceManager.cs (hunk 8 above)
Comment (MikhailArkhipov) [inline src/Common/Core/Impl/Services/ServiceManager.cs]: `VsAppShellTestSetup`

Comment (MikhailArkhipov) [inline src/Common/Core/Impl/Services/ServiceManager.cs]: Guess need to add tests

**Pull request merged**